### ISSUE TITLE
TypeVar with bound union fails to narrow type

### ISSUE THREAD

**Issue by Dreamsorcerer** [python/mypy#13896]
```
from functools import wraps
from typing import Awaitable, Callable, Coroutine, TypeVar
from aiohttp import web
from aiojobs.aiohttp import spawn

_RequestView = TypeVar("_RequestView", bound=web.Request | web.View)
_T = TypeVar("_T")

def atomic(
    coro: Callable[[_RequestView], Coroutine[object, object, _T]]
) -> Callable[[_RequestView], Awaitable[_T]]:
    @wraps(coro)
    async def wrapper(request_or_view: _RequestView) -> _T:
        if isinstance(request_or_view, web.View):
            # Class Based View decorated.
            request = request_or_view.request
        else:
            request = request_or_view

        job = await spawn(request, coro(request_or_view))
        return await job.wait()

    return wrapper

class MyView(web.View):
    @atomic
    async def get(self) -> web.Response:
        return web.Response(text=self.request.method)
```
This gives:
```
error: Incompatible types in assignment (expression has type "_RequestView", variable has type "Request")  [assignment]
```

Type narrowing seems to work when doing `TypeVar("_RequestView", web.Request. web.View)`.

Full code: https://github.com/aio-libs/aiojobs/pull/352/files#diff-9d3ec99bcf5007c941d322795c930bcd9a430db47d2416d6146f14b8c9db66c2R37

**Comment by JelleZijlstra** [comment]
How about defining the TypeVar with `bound=web.Request | web.View`?

**Comment by Dreamsorcerer** [comment]
Yep, so that seems to be mostly there (and seems to fix #13895, which I was encountering at the same time). Only issue is that it seems to fail type narrowing on the `else`, so I end up with:
```
aiojobs/aiohttp.py:46:23: error: Incompatible types in assignment (expression has type "_RequestView", variable has type "Request")  [assignment]
                request = request_or_view
                          ^~~~~~~~~~~~~~~
```

So, the type is narrowed in the else to `web.View`, but is not narrowed in the else, which seems odd to me.

**Comment by Dreamsorcerer** [comment]
Seems to be fixed in 1.12:
https://github.com/aio-libs/aiojobs/pull/520
